### PR TITLE
Two I2C RTC modules

### PR DESCRIPTION
**Pull request by jipanienko**
Not relay sure if may be usefully to other but I share this.

Two RTC module with same functiones that can keep wall clock

### CODE REVIEW

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    assert(rc == 0);
+}
+
+struct ds1307_dev * ds1307_open(const char *name, struct ds1307_cfg *cfg)
+{
+    return (struct ds1307_dev *)os_dev_open(name, 0, cfg);
+}
+
+struct ds1307_dev ds1307;
+struct ds1307_hw_cfg hw_cfg = {
```
Comment (sjanc) [inline hw/drivers/rtc/ds1307/src/ds1307.c:241]: those could ne static

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+void ds1307_pkg_init(void)
+{
+    struct ds1307_dev *rtc;
+    struct clocktime ct;
+    struct os_timeval tv;
+    struct os_timezone tz;
+
+    ds1307_create(&ds1307, MYNEWT_VAL(DS1307_OS_DEV_NAME), &hw_cfg);
+#if 1
```
Comment (sjanc) [inline hw/drivers/rtc/ds1307/src/ds1307.c:253]: some leftover ?

**Review by sjanc — APPROVED** [review]
looks good to me, some minor comments only

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+    struct hal_i2c_master_data data_struct = {
+        .address = DS1307_I2C_ADDR,
+        .len = count + 1,
+        .buffer = payload,
+    };
+
+    STATS_INC(ds1307->stats, write_count);
+    rc = i2cn_master_write(ds1307->hw_cfg.i2c_num, &data_struct, OS_TICKS_PER_SEC / 10, 1, 1);
+    if (rc)
```
Comment (kasjer) [inline hw/drivers/rtc/ds1307/src/ds1307.c:52]: mynewt uses `{` in same line as `if`, I don't know why CI did not detect this automatically @utzig

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        STATS_INC(ds1307->stats, read_errors);
+        DS1307_LOG_ERROR("DS1307 write I2C failed\n");
+        goto exit;
+    }
+
+    data_struct.len = count;
+    data_struct.buffer = regs;
+    rc = i2cn_master_read(ds1307->hw_cfg.i2c_num, &data_struct, OS_TICKS_PER_SEC / 10, 1, 1);
+    if (rc)
+    {
```
Comment (kasjer) [inline hw/drivers/rtc/ds1307/src/ds1307.c:84]: move { to previous line

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        STATS_INC(ds1307->stats, read_errors);
+        DS1307_LOG_ERROR("DS1307 read I2C failed\n");
+    }
+
+exit:
+    return rc;
+}
+
+static int bcd_to_bin(uint8_t bcd)
```
Comment (kasjer) [inline hw/drivers/rtc/ds1307/src/ds1307.c:93]: in this case line should be split after int

File: hw/drivers/rtc/ds1307/src/ds1307.c [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+int ds1307_read_time(struct ds1307_dev *ds1307, struct clocktime *tm)
+{
+    int rc;
+    uint8_t buf[0x13];
+
+    rc = ds1307_read_regs(ds1307, DS1307_SECONDS_ADDR, buf, 0x13);
+    if (rc == 0)
+    {
```
Comment (kasjer) [inline hw/drivers/rtc/ds1307/src/ds1307.c:105]: move { to previous line

File: hw/drivers/rtc/ds3231/include/ds3231/ds3231.h [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ *
+ * @param dev  Pointer to the ds3231_dev device descriptor
+ * @param arg  Pointer to struct ds3231_hw_cfg.
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int ds3231_init(struct os_dev *dev, void *arg);
+
+/**
+ * Sets sensor device configuration.
```
Comment (kasjer) [inline hw/drivers/rtc/ds3231/include/ds3231/ds3231.h:106]: I guess this is not really a sensor (copy/paste from sensor code?)

File: hw/drivers/rtc/ds3231/src/ds3231_shell.c [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+static int ds3231_shell_err_unknown_arg(char *cmd_name)
+{
+    console_printf("Error: unknown argument \"%s\"\n",
+                   cmd_name);
+    return EINVAL;
+}
+
+int ds3231_shell_err_invalid_arg(char *cmd_name)
```
Comment (kasjer) [inline hw/drivers/rtc/ds3231/src/ds3231_shell.c:48]: could be static and int should be in separate line

File: hw/drivers/rtc/ds3231/src/ds3231.c [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+void ds3231_pkg_init(void)
+{
+    struct ds3231_dev *rtc;
+    struct clocktime ct;
+    struct os_timeval tv;
+    struct os_timezone tz;
+
+    ds3231_create(&ds3231, MYNEWT_VAL(DS3231_OS_DEV_NAME), &hw_cfg);
+#if 1
```
Comment (kasjer) [inline hw/drivers/rtc/ds3231/src/ds3231.c:269]: some left over

File: hw/drivers/rtc/ds3231/include/ds3231/ds3231.h [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+ * @param  ds3231      Pointer to the ds3231 struct
+ * @param  ds3231_cfg  Pointer to the ds3231 configuration
+ *
+ * @return 0 on success, non-zero on failure.
+ */
+int ds3231_config(struct ds3231_dev *ds3231, const struct ds3231_cfg *cfg);
+
+int ds3231_read_time(struct ds3231_dev *ds3231, struct clocktime *tm);
+
+int ds3231_read_temp(struct ds3231_dev *ds3231, int16_t *temperature);
```
Comment (kasjer) [inline hw/drivers/rtc/ds3231/include/ds3231/ds3231.h:117]: maybe a comment here what temperature unit is?

File: hw/drivers/rtc/ds1307/src/ds1307_shell.c [hunk 11]
```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2020 Jesus Ipanienko
```
Comment (kasjer) [inline hw/drivers/rtc/ds1307/src/ds1307_shell.c:2]: This file seems to be older then others :)

File: hw/drivers/rtc/ds3231/src/ds3231_shell.c [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+ * limitations under the License.
+ */
+
+#include <string.h>
+#include <errno.h>
+#include "os/mynewt.h"
+#include "console/console.h"
+#include "shell/shell.h"
+#include "ds3231/ds3231.h"
+#include "parse/parse.h"
```
Comment (kasjer) [inline hw/drivers/rtc/ds3231/src/ds3231_shell.c:23]: slight inconsistency `<` `>` are use in all other files

**Review by kasjer — COMMENTED** [review]
Overall looks fine

**Review by kasjer — APPROVED** [review]
Looks goot

**Pull request merged**